File: 823-binary-trees-with-factors/823-binary-trees-with-factors.cpp
#define ll long long int
class Solution {
    const int M=1e9+7;
    map<int,ll>BTs;
public:
    int numFactoredBinaryTrees(vector<int>& arr) 
    {
        sort(arr.begin(),arr.end());
        
        BTs[arr[0]]=1LL;
        int n=arr.size();
        for(int i=1;i<n;i++)
        {
            BTs[arr[i]]=1LL;
            for(int j=0;1LL*arr[j]*arr[j]<=1LL*arr[i];j++)
            {
                if(arr[i]%arr[j]==0&&BTs.count(arr[i]/arr[j]))
                {
                    if(arr[j]*arr[j]!=arr[i])
                    BTs[arr[i]]=BTs[arr[i]]%M+(2LL*BTs[arr[j]]%M*BTs[arr[i]/arr[j]]%M)%M;
                    else
                        BTs[arr[i]]=BTs[arr[i]]%M+(1LL*BTs[arr[j]]%M*BTs[arr[i]/arr[j]]%M)%M;
                }
            }
        }
        
        ll res=0;
        for(int i=0;i<n;i++)
           res=(res%M+BTs[arr[i]]%M)%M;
        
        return res;
    }
};

File: 823-binary-trees-with-factors/823-binary-trees-with-factors.java
class Solution 
{
    int mod=(int)Math.pow(10,9)+7;
    
    public int numFactoredBinaryTrees(int[] arr) 
    {
        int n=arr.length;
        long ans=0;
        
        Arrays.sort(arr);
        
        long count[]=new long[n];
        count[0]=1;ans++;
        
        for(int i=1;i<n;i++)
        {
            count[i]=1;
            Map<Integer,Integer> map=new HashMap<>();
            
            for(int j=0;j<i;j++)
            {
                if(arr[i]%arr[j]==0)
                {
                    map.put(arr[j],j);
                    
                    if(map.containsKey(arr[i]/arr[j]))
                    {
                        long x=count[j]%mod;
                        long y=count[map.get(arr[i]/arr[j])]%mod;
                        
                        long prod=(x*y)%mod;
                        count[i]=((count[i]%mod)+(prod%mod))%mod;
                        
                        if(arr[i]/arr[j]!=arr[j])
                        count[i]=((count[i]%mod)+(prod%mod))%mod;
                    }
                }
            }
            ans=((ans%mod)+(count[i]%mod))%mod;
        }
        
        return (int)(ans%mod);
        
    }
}